### DISCUSSION TITLE
[Application Development] Gas oracle returns 0 base fees

### DISCUSSION THREAD

**Discussion by andy-uphold** [ethereum-optimism/developers#247] [Application Development] [answered]
### Discussed in https://github.com/ethereum-optimism/developers/discussions/224

<div type='discussions-op-text'>

<sup>Originally posted by **bejitono** February  1, 2024</sup>
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

We are doing an eth_call to read the `baseFee` from the gas oracle contract `0x420000000000000000000000000000000000000F` which has worked fine until today but is now returning a base fee of 0. We're doing the same on Base which currently works fine.

### Additional Information

Example request: 
`{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f"},"latest"],"jsonrpc":"2.0","id":1}`

### Feedback

_No response_</div>

**Comment by andy-uphold** [comment]
Can we get some more information on the real issue here? 

Making a call on a contract that requires passing in gas information is VERY strange.

I find it hard to believe that geth is to blame for this - can you show us how geth is enforcing this? - as this would severely limit how calls are made. 

From experimenting with this, because you pass in the maxFeePerGas, the call is now requiring that you actually have enough fees to pay for the transaction, which is very frustrating.

I have also noticed that this is only a problem when I have made the call via alchemy or mainnet.optimism.io endpoints, whereas:

https://optimistic.etherscan.io/address/0x420000000000000000000000000000000000000f#readProxyContract

is fine.

**Comment by sbvegan** [comment, marked as answer]
It's not really an "issue", it's a change in behavior of upstream geth's rpc api and a new quirk that JSON-RPC API callers need to handle. This is the context I got from client engineers:

> Yeah I think this is actually correct behaviour from geth - executing a transaction with 0 gas price as you’re trying to do in eth_call will effect the EVM context being created because it’s not a realistic situation post EIP1559. You can imagine some contracts subtracting the base fee from the gas price to find the inclusion fee that would hit underflow errors if the base fee wasn’t set to zero when there’s no gas price.

Here's the base fee being explicitly set to `0` when using `eth_call` in our code: 

https://github.com/ethereum-optimism/op-geth/blob/ea06bb79963e6fd091ff6e1b8b24e7e4355ff334/core/vm/evm.go#L135-L142

where NoBaseFee gets set:

https://github.com/ethereum-optimism/op-geth/blob/75ee01e842954cfacc19aff7061ece6960fd15a1/internal/ethapi/api.go#L1175

and here's the upstream geth code:

https://github.com/ethereum-optimism/op-geth/commit/470dba8fc1890938a65bbf4293a4759a9b9615a1

## Edit - Additional Context

If basefee tracking is disabled (eth_call, eth_estimateGas, etc), and no gas prices were specified, lower the basefee to 0 to avoid breaking EVM. So when using `eth_call` to this contract the value sets the value to zero. Read the code above for full context. You can adjust by making a request with a `gasPrice` parameter set. For example:

```
curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f", "gasPrice":"0x100000000"},"latest"],"jsonrpc":"2.0","id":1}' | jq
```

For estimating gas you can also use the tools that are outlined on our documentation [here](https://docs.optimism.io/builders/dapp-developers/transactions/estimates#tooling). Using a developer tool will help reduce the confusion of fee calculation by abstracting the details away. It has the added bonus of working out of the box with the Ecotone upgrade (blobs) by simply updating to a newer version of the tool.

**Reply by 0xtakamaka** [reply]
How can we (JSON-RPC API callers) handle 0 baseFee? When we calculate required gas prices (l1BaseFee and gasPrice or baseFee) from contract 0x420000000000000000000000000000000000000F, we need baseFee data, to properly calculate required fee, as described in https://docs.optimism.io/builders/dapp-developers/transactions/fees.

How can we calculate this if baseFee on contract 0x420000000000000000000000000000000000000F returns 0?

**Reply by sbvegan** [reply]
You can use a request like this:

```
curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f", "gasPrice":"0x100000000"},"latest"],"jsonrpc":"2.0","id":1}' | jq
```

Or you can use existing tooling outlined here: https://docs.optimism.io/builders/dapp-developers/transactions/estimates#tooling

**Comment by andy-uphold** [comment]
Thanks for the reply :)

Don't you think it's an issue when a user has to have a certain amount of funds in their account in order to call a read-only method on a contract?

If this is due to an Optimism specific fork of geth, then this should be fixed there, right?

**Comment by andy-uphold** [comment]
The issue really is that you are asking people to set a value in their request which is at least as big as the value that they want to know. 

This is especially an issue as passing in an arbitrarily large value (ideally max uint256 to cover all eventualities) to get the lower precise value is not possible due to needing funds.

**Comment by sbvegan** [comment]
What funding do you need? These are just requests that I've made from my terminal. Its not connected to an account at all. `op-geth` is a [minimal diff](https://op-geth.optimism.io/) from `geth` and we aim to keep the upstream code mostly the same with our implementation.

## Your Request

```
$ curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f"},"latest"],"jsonrpc":"2.0","id":1}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   239  100   103  100   136    394    520 --:--:-- --:--:-- --:--:--   919
{
  "jsonrpc": "2.0",
  "result": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "id": 1
}
```

## Request with small `gasPrice`

```
$ curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f", "gasPrice":"0x1"},"latest"],"jsonrpc":"2.0","id":1}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   373  100   219  100   154    674    473 --:--:-- --:--:-- --:--:--  1151
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32000,
    "message": "err: max fee per gas less than block base fee: address 0x0000000000000000000000000000000000000000, maxFeePerGas: 1, baseFee: 99969359 (supplied gas 50000000)"
  },
  "id": 1
}
```

## Request with `gasPrice`

```
curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f", "gasPrice":"0x100000000"},"latest"],"jsonrpc":"2.0","id":1}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   265  100   103  100   162    368    579 --:--:-- --:--:-- --:--:--   949
{
  "jsonrpc": "2.0",
  "result": "0x0000000000000000000000000000000000000000000000000000000005fdb229",
  "id": 1
}
```

**Reply by andy-uphold** [reply]
% curl https://mainnet.optimism.io \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"eth_call","params":[{"data":"0x6ef25c3a","to":"0x420000000000000000000000000000000000000f", "gasPrice":"0xF00000000"},"latest"],"jsonrpc":"2.0","id":1}'
{"jsonrpc":"2.0","error":{"code":-32000,"message":"err: insufficient funds for gas * price + value: address 0x0000000000000000000000000000000000000000 have 1026258964425397610 want 3221225472000000000 (supplied gas 50000000)"},"id":1}

**Reply by sbvegan** [reply]
You can just remove a `0` from your `gasPrice` and then the query works
